Revamp ClusterWorkloadResourceMapping (#183)
servicebinding/spec: README.md

[cwrm-crd]: servicebinding.io_clusterworkloadresourcemappings.yaml
[container]: https://kubernetes.io/docs/reference/kubernetes-api/workloads-resources/container/
[crd-syntax]: https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#create-a-customresourcedefinition
[envvar]: https://kubernetes.io/docs/reference/kubernetes-api/workloads-resources/container/#environment-variables

> [jsonpath]: http://goessner.net/articles/JsonPath/

Author: I didn't want to link to a document that expires. The content is functional equivalent to http://goessner.net/articles/JsonPath/ (which is the authoritative source for JSONPath until the rfc is active)

[volume]: https://kubernetes.io/docs/reference/kubernetes-api/config-and-storage-resources/volume
[volumemount]: https://kubernetes.io/docs/reference/kubernetes-api/workloads-resources/container/#volumes

## Restricted JSONPath

> > Note: only expression labels as 'Restricted JSONPath' **MUST** conform to this requirement. Other expressions **MAY** use the full JSONPath syntax.

Reviewer: ```suggestion
> Note: Only expressions labeled as 'Restricted JSONPath' **MUST** conform to this requirement. Other expressions **MAY** use the full JSONPath syntax.
```

Reviewer: Since this is a normative text, I think we can avoid `Note:`.

Author: ```suggestion
> only expression labels as 'Restricted JSONPath' **MUST** conform to this requirement. Other expressions **MAY** use the full JSONPath syntax.
```

> A Restricted JSONPath is a subset of [JSONPath][jsonpath] expressions that **MUST NOT** use type and operators other than fields separated by the child operator.

Author: I agree for a user guide but not for the spec. We're not defining a floor of functionality in this case we're defining the ceiling.

> A Restricted JSONPath is a subset of [JSONPath][jsonpath] expressions that **MUST NOT** use type and operators other than fields separated by the child operator.

Reviewer 2: well, but json path is already defined as it is and there is no movement on that front. Hence, it is clearer to state here what subset of the the jsonpath spec we understand as "restricted jsonpath", Otherwise, we are forcing the developers to figure that out by themselves.

> A Restricted JSONPath is a subset of [JSONPath][jsonpath] expressions that **MUST NOT** use type and operators other than fields separated by the child operator.

Author: I think the current text says what your asking for. Would you like to propose alternate language?

> A Restricted JSONPath is a subset of [JSONPath][jsonpath] expressions that **MUST NOT** use type and operators other than fields separated by the child operator.

Reviewer 2: Yes, I would like to propose that we say something like this:

A Restricted JSONPath is a subset of [JSONPath][jsonpath] expressions that **MUST** consists of dot-selectors only.

```
restricted-json-path = +(dot-selector)
dot-selector    = "." dot-member-name
   dot-member-name = name-first *name-char
   name-first =
                         ALPHA /
                         "_"   /           ; _
                         %x80-10FFFF       ; any non-ASCII Unicode character
   name-char = DIGIT / name-first

   DIGIT           =  %x30-39              ; 0-9
   ALPHA           =  %x41-5A / %x61-7A    ; A-Z / a-z   
```

IMHO, this describe exactly what is allowed by the spec.

> A Restricted JSONPath is a subset of [JSONPath][jsonpath] expressions that **MUST NOT** use type and operators other than fields separated by the child operator.

Author: "selectors" are introduced in the rfc, it's strange to use them for our definition which isn't based on the rfc. Moreover, the dot selector does not fully cover the "child operator" which allows for field access via dot or square bracket notation.

I'm going to keep the current text instead of your proposal. The "child operator" is defined and the examples below this line clearly show what is allowed. We can revisit this once the rfc is no longer a draft and we switch all JSONPaths over to the rfc's definition.

  versions:
  - version: "*"
    containers:
    - path: .spec.template.spec.containers[*]
      name: /name
      env: /env
      volumeMounts: /volumeMounts
    - path: .spec.template.spec.initContainers[*]
      name: /name
      env: /env
      volumeMounts: /volumeMounts
    volumes: /spec/template/spec/volumes
```

Note: this example is equivalent to not specifying a mapping, or specifying an empty mapping.

## Reconciler Implementation

A reconciler implementation **MUST** support mapping to PodSpec-able resources without defining a `ClusterWorkloadResourceMapping` for those types.

> If a `ServiceBinding` specifies `.spec.workload.containers` and a `MappingContainer` specifies a `name` pointer, the resolved name **MUST** be used to filter container based on the.

Author: fixed

  versions:
  - version: "*"
    containers:
    - path: .spec.template.spec.containers[*]
      name: .name
      env: .env
      volumeMounts: .volumeMounts
    - path: .spec.template.spec.initContainers[*]
      name: .name
      env: .env
      volumeMounts: .volumeMounts

>     volumes: .spec/template.spec.volumes

Reviewer 3: @scothis - did we miss a `/` replacement here?

Author: ```suggestion
    volumes: .spec.template.spec.volumes
```

```

> Note: this example is equivalent to not specifying a mapping, or specifying an empty mapping.

Reviewer: ```suggestion
Note: This example is equivalent to not specifying a mapping or specifying an empty mapping.
```

## Reconciler Implementation

> A reconciler implementation **MUST** support mapping to PodSpec-able resources without defining a `ClusterWorkloadResourceMapping` for those types.

Reviewer 2: Does this mean that a reconciler MUST be able to detect if resource is PodSpec-able? Or there is a simple a list of PodSpec-able resources that are supported without adding  `ClusterWorkloadResourceMapping` ?

Author: Neither. It's saying that unless a ClusterWorkloadResourceMapping is define, treat the workload resource as if it was PodSpecable. If it's not PodSpecable I'd expect validation to fail at admission time.

Reviewer 2: > unless a ClusterWorkloadResourceMapping is defined, treat the workload resource as if it was PodSpecable. If it's not PodSpecable I'd expect validation to fail at admission time

Having the above sentence in the spec text would be very helpful.

Author: refined the text, take a look

> If a `ServiceBinding` specifies `.spec.workload.containers` and a `MappingContainer` specifies a `name` expression, the resolved name **MUST** limit which containers in the workload are bound.

Reviewer 2: What happens when `name` expression is not specified, but `.spec.workload.containers` is there?

Author: In that case there is no requirement to restrict binding to the containers specified by `.spec.workload.containers`. The other option would be to fail loudly.

Reviewer 2: I would vote for later, and state that in the spec.

Author: I'm not convinced erroring is the best behavior in all cases. Consider the RuntimeComponent resource where there's and unnamed implicit container at the root of the spec and sets of named init/sidecar container, if we fail hard no filtering of the init and sidecar containers would be possible.

Author: I added a suggestion that if either field is missing, that the container should be bound.

> A reconciler **MUST** create empty values at locations referenced by Restricted JSONPaths that do not exist on the workload resource. Values referenced by JSONPaths in both the `MappingTemplate` and `MappingContainer`s **MUST** be mutated by a `ServiceBinding` reconciler as if they were defined directly by a PodTemplateSpec. A reconciler **MUST** preserve fields on the workload resource that fall outside the specific fragments and types defined by the mapping.

Reviewer: ```suggestion
A reconciler **MUST** create empty values at locations referenced by [Restricted JSONPaths](#restricted-jsonpath) that do not exist on the workload resource. Values referenced by JSONPaths in both the `MappingTemplate` and `MappingContainer`s **MUST** be mutated by a `ServiceBinding` reconciler as if they were defined directly by a PodTemplateSpec. A reconciler **MUST** preserve fields on the workload resource that fall outside the specific fragments and types defined by the mapping.
```
